### ISSUE TITLE
take_while_m_n fails on multi-byte characters

### ISSUE THREAD

**Issue by Recmo** [Geal/nom#913]
## Prerequisites

- Rust version : `rustc 1.34.0 (91856ed52 2019-04-10)`
- nom version : `4.2.2`
- nom compilation features used: default

## Test case

Please provide a short, complete (with crate import, etc) test case for
the issue, showing clearly the expected and obtained results.

Example test case:

```rust
use nom::*;
named!(parser<&str, &str>, take_while_m_n!(1, 1, |c| c == 'A' || c == '😃'));

fn main() {
    assert_eq!(parser("A!"), Ok(("!", "A")));
    assert_eq!(parser("😃!"), Ok(("!", "😃")));
}
```

```
thread 'main' panicked at 'byte index 1 is not a char boundary; it is inside '😃' (bytes 0..4) of `😃!`', src/libcore/str/mod.rs:2027:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:70
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:58
             at src/libstd/panicking.rs:200
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:215
   4: <std::panicking::begin_panic::PanicPayload<A> as core::panic::BoxMeUp>::get
             at src/libstd/panicking.rs:478
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:385
   6: std::panicking::try::do_call
             at src/libstd/panicking.rs:312
   7: <T as core::any::Any>::type_id
             at src/libcore/panicking.rs:85
   8: core::str::traits::<impl core::slice::SliceIndex<str> for core::ops::range::Range<usize>>::index::{{closure}}
             at src/libcore/str/mod.rs:2027
   9: core::str::<impl str>::as_ptr
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libcore/str/mod.rs:1877
  10: <core::result::Result<T, E> as core::cmp::PartialEq>::eq
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libcore/option.rs:386
  11: core::str::traits::<impl core::slice::SliceIndex<str> for core::ops::range::RangeFrom<usize>>::index
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libcore/str/mod.rs:1877
  12: core::str::<impl str>::as_ptr
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libcore/str/mod.rs:1616
  13: <&'a str as nom::traits::InputTake>::take_split
             at /Users/remco/.cargo/registry/src/github.com-1ecc6299db9ec823/nom-4.2.2/src/traits.rs:448
  14: nom_bug::parser
             at /Users/remco/WF/olus-lang/olus-rust/<::nom::macros::named macros>:13
  15: nom_bug::main
             at nom-bug/src/main.rs:5
  16: std::rt::lang_start::{{closure}}
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libstd/rt.rs:64
  17: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:297
  18: panic_unwind::dwarf::eh::read_encoded_pointer
             at src/libpanic_unwind/lib.rs:87
  19: <std::panicking::begin_panic::PanicPayload<A> as core::panic::BoxMeUp>::get
             at src/libstd/panicking.rs:276
             at src/libstd/panic.rs:388
             at src/libstd/rt.rs:48
  20: std::rt::lang_start
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libstd/rt.rs:64
  21: nom_bug::parser
```

**Comment by Recmo** [comment]
This test case is artificial, it can be done using `char!` and then all works fine. But for my usecase I need a predicate on a char (since I will look up character properties in the unicode database).

I think the problem is that `take_while_m_n!` takes it's arguments in bytes, but the predicate is in codepoints.

**Comment by Geal** [comment]
could you test with the version on master?

**Comment by Geal** [comment]
this should be fixed by 7483885 and will be available in nom 5
